### PR TITLE
add more method for timer output

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR adds more statistical methods like average, median, best, and worst to compute the time result.

When the timing use x iteration which is more than 1 iteration, timer will record the average time x times.
Handle repetition_growth_factor <=1 case such that next_timing always be next iteration.
User can set it to get the same overhead operation in each repetitions

### CODE REVIEW

File: benchmark/utils/general.hpp [hunk 1]
```diff
@@ -581,10 +581,27 @@ class IterationControl {
...
+     * @param method  the statistical method
+     * @param same_rep  whether timer managed outside has the same repetitions
+     *                  as IterationControl. If the timer is managed outside and
+     *                  the timer does not have the same repetitions as
+     *                  InterationControl, the method is only allowed "average".
+     *
+     * @return the statistical time
+     */
+    double compute_time(const std::string& method = "average",
+                        bool same_rep = true) const
```
Comment (pratikvn) [inline benchmark/utils/general.hpp:596]: Do you mean to set this to `false`  ?

File: benchmark/utils/timer_impl.hpp [hunk 2]
```diff
@@ -63,13 +63,18 @@ class Timer {
 
     /**
      * Finish the timer
+     *
+     * @param num  the number of operation for the timing range
```
Comment (yanjen) [inline benchmark/utils/timer_impl.hpp:67]: "the number of samples for the statistical timing" might be better?

File: benchmark/utils/general.hpp [hunk 3]
```diff
@@ -581,10 +581,25 @@ class IterationControl {
...
+            this->get_num_repetitions()) {
+            return status_run_.managed_timer.compute_time(method);
+        } else {
+            assert(method == "average");
+            return status_run_.managed_timer.get_total_time() /
+                   this->get_num_repetitions();
+        }
     }
 
     IndexType get_num_repetitions() const { return status_run_.cur_it; }
```
Comment (yanjen) [inline benchmark/utils/general.hpp:605]: It was not from this pr but I think the word "samples" would describe better than the word "repetitions". Does anyone feel the same way?

File: benchmark/blas/blas_common.hpp [hunk 4]
```diff
@@ -425,7 +425,7 @@ void apply_blas(const char* operation_name, std::shared_ptr<gko::Executor> exec,
         for (auto _ : ic.run()) {
             op->run();
         }
-        const auto runtime = ic.compute_average_time();
+        const auto runtime = ic.compute_time(FLAGS_timer_output);
```
Comment (yanjen) [inline benchmark/blas/blas_common.hpp:428]: FLAGS_timer_type or might be better? Same for the other places.

**Review by yanjen — APPROVED** [review]
LGTM, but I have some suggestions on the naming.

**Review by MarcelKoch — COMMENTED** [review]
I'm not sure if this makes sense. Currently, we are not timing each iteration individually, in order to reduce the timing overhead. Perhaps it makes sense to revisit that. But for now, without timings of the individual iterations, it doesn't make sense to compute the median, or min/max.

**Review by MarcelKoch — COMMENTED** [review]
Currently, we take timestamps after a varying number of iterations, for example at repetition `1, 2, 4, 8, 16, 32, ...` (not exactly but similar). So with the proposed changes we would take averages of a growing number of repetitions. That might have useful information, especially considering that the kernel launch and synchronization overhead is larger for fewer repetitions. But we wouldn't provide actual medians, or min/max.
Still, I won't hold this up.

File: benchmark/utils/general.hpp (hunk 3 above)
Comment (yhmtsai) [reply benchmark/utils/general.hpp:605]: I think the repetitions is better. Sample to me is like subset of data.

File: benchmark/blas/blas_common.hpp (hunk 4 above)
Comment (yhmtsai) [reply benchmark/blas/blas_common.hpp]: I change it to timer_method

**Comment by yhmtsai** [comment]
rebase!

File: benchmark/utils/general.hpp [hunk 5]
```diff
@@ -115,9 +115,10 @@ DEFINE_uint32(max_repetitions, std::numeric_limits<unsigned int>::max(),
               "If 'repetitions = auto' is used, the maximal number of"
               " repetitions for a single benchmark.");
 
-DEFINE_double(repetition_growth_factor, 1.5,
-              "If 'repetitions = auto' is used, the factor with which the"
-              " repetitions between two timings increase.");
+DEFINE_double(
+    repetition_growth_factor, 1.5,
+    "The factor with which the repetitions between two timings increase. If it "
+    "is lower than or equal to 1, the timing region always be 1 repetition.");
```
Comment (upsj) [inline benchmark/utils/general.hpp:121]:
```suggestion
    "is lower than or equal to 1, the timing region is always 1 repetition.");
```

File: benchmark/utils/timer_impl.hpp [hunk 6]
```diff
@@ -87,13 +92,31 @@ class Timer {
...
-        return this->get_total_time() / this->get_num_repetitions();
+        if (method == "average") {
+            return this->get_total_time() / this->get_num_repetitions();
+        }
+        auto copy = duration_sec_;
+        std::sort(copy.begin(), copy.end());
+        if (method == "best") {
+            return copy.front();
+        } else if (method == "worst") {
+            return copy.back();
```
Comment (upsj) [inline benchmark/utils/timer_impl.hpp:111]: how about using min and max? For times, this should be unambiguous and is closer to what is used in statistical analysis

**Pull request merged**